apps/src/lib/node/ledger/protocol/transactions/validator_set_update/mod.rs

```diff
@@ -31,29 +34,116 @@ where
     D: 'static + DB + for<'iter> DBIter<'iter> + Sync,
     H: 'static + StorageHasher + Sync,
 {
+    if ext.signatures.is_empty() {
+        tracing::debug!("Ignoring empty validator set update");
+        return Ok(Default::default());
+    }
+
     tracing::info!(
         num_votes = ext.signatures.len(),
         "Aggregating new votes for validator set update"
     );
 
     let voting_powers = utils::get_voting_powers(storage, &ext)?;
-    let changed_keys = apply_updates(storage, ext, voting_powers)?;
+    let changed_keys = apply_update(storage, ext, voting_powers)?;
 
     Ok(TxResult {
         changed_keys,
         ..Default::default()
     })
 }
 
-fn apply_updates<D, H>(
-    _storage: &mut Storage<D, H>,
-    _ext: validator_set_update::VextDigest,
-    _voting_powers: HashMap<(Address, BlockHeight), FractionalVotingPower>,
+fn apply_update<D, H>(
+    storage: &mut Storage<D, H>,
+    ext: validator_set_update::VextDigest,
+    voting_powers: HashMap<(Address, BlockHeight), FractionalVotingPower>,
 ) -> Result<ChangedKeys>
 where
     D: 'static + DB + for<'iter> DBIter<'iter> + Sync,
     H: 'static + StorageHasher + Sync,
 {
-    tracing::warn!("Called apply_updates() with no side effects");
-    Ok(ChangedKeys::new())
+    let epoch = {
+        // all votes we gathered are for the same epoch, so
+        // we can just fetch the block height from the first
+        // signature we iterate over, and calculate its cor-
+        // responding epoch
+        let height = ext
+            .signatures
+            .keys()
+            .map(|(_, height)| *height)
+            .by_ref()
```

Comment:
do we need this `.by_ref()`?

Reply:
probably not. it creates a `&mut` reference. I thought we would need it to manually call `next()` on the iterator, without previously giving it some `let` binding